### PR TITLE
[Major] Support Custom Learning Rate Scheduler

### PR DESCRIPTION
**Pull request by ourownstory**
Changes:
* Support any learning rate scheduler
* quantiles stored in model_config instead of train_config
* improve training progress tracking in model
* pass config_model to TimeNet
* introduce finding_lr flag in TimeNet
* call self.config_train.set_optimizer and self.config_train.set_scheduler in TimeNet configure_optmizer
* remove train_loader from self in TimeNet
* create utils_lightning file with find_learning_rate and configure_tuner
* update test to only invoke lr_finder explicilty
* create test_save and test_train_config files

Future TODOs: 
* improve scheduler defaults
* improve what gets logged and progress printing when finding_lr

### CODE REVIEW

File: neuralprophet/time_net.py [hunk 1]
```diff
@@ -861,48 +868,59 @@ def predict_step(self, batch, batch_idx, dataloader_idx=0):
...
+        self._scheduler = self.config_train.scheduler
+
         # Optimizer
         optimizer = self._optimizer(self.parameters(), lr=self.learning_rate, **self.config_train.optimizer_args)
 
         # Scheduler
-        lr_scheduler = self._scheduler(
-            optimizer,
-            max_lr=self.learning_rate,
-            total_steps=self.trainer.estimated_stepping_batches,
```
Comment (ourownstory) [inline neuralprophet/time_net.py:871]: redo this

File: neuralprophet/time_net.py [hunk 2]
```diff
@@ -861,48 +868,59 @@ def predict_step(self, batch, batch_idx, dataloader_idx=0):
...
+            )
+        else:
+            lr_scheduler = self._scheduler(
+                optimizer,
+                **self.config_train.scheduler_args,
+            )
 
         return {"optimizer": optimizer, "lr_scheduler": lr_scheduler}
 
     def _get_time_based_sample_weight(self, t):
```
Comment (ourownstory) [inline neuralprophet/time_net.py:898]: undo this

File: neuralprophet/time_net.py [hunk 3]
```diff
@@ -794,7 +801,7 @@ def training_step(self, batch, batch_idx):
         optimizer.step()
 
         scheduler = self.lr_schedulers()
-        scheduler.step()
+        scheduler.step(epoch=self.train_progress)
```
Comment (ourownstory) [inline neuralprophet/time_net.py:804]: undo this

File: neuralprophet/time_net.py [hunk 4]
```diff
@@ -761,20 +762,21 @@ def loss_func(self, inputs, predicted, targets):
...
         # Compute loss. no reduction.
         loss = self.config_train.loss_func(predicted, targets)
-        # Weigh newer samples more.
-        loss = loss * self._get_time_based_sample_weight(t=inputs["time"][:, self.n_lags :])
+        if self.config_train.newer_samples_weight > 1.0:
+            # Weigh newer samples more.
+            loss = loss * self._get_time_based_sample_weight(t=inputs["time"][:, self.n_lags :])
         loss = loss.sum(dim=2).mean()
         # Regularize.
         if self.reg_enabled:
```
Comment (ourownstory) [inline neuralprophet/time_net.py:770]: disable when finding_lr

File: neuralprophet/time_net.py [hunk 5]
```diff
@@ -794,19 +796,25 @@ def training_step(self, batch, batch_idx):
...
+        else:
+            scheduler.step(epoch=epoch_float)
 
-        # Manually track the loss for the lr finder
-        self.log("train_loss", loss, on_step=False, on_epoch=True, prog_bar=True, logger=True)
-        self.log("reg_loss", reg_loss, on_step=False, on_epoch=True, prog_bar=True, logger=True)
+        if self.finding_lr:
+            # Manually track the loss for the lr finder
+            self.log("train_loss", loss, on_step=False, on_epoch=True, prog_bar=True, logger=True)
+            self.log("reg_loss", reg_loss, on_step=False, on_epoch=True, prog_bar=True, logger=True)
```
Comment (ourownstory) [inline neuralprophet/time_net.py:807]: skip reg_loss

File: neuralprophet/time_net.py [hunk 6]
```diff
@@ -794,19 +796,25 @@ def training_step(self, batch, batch_idx):
...
+            scheduler.step()
+        else:
+            scheduler.step(epoch=epoch_float)
 
-        # Manually track the loss for the lr finder
-        self.log("train_loss", loss, on_step=False, on_epoch=True, prog_bar=True, logger=True)
-        self.log("reg_loss", reg_loss, on_step=False, on_epoch=True, prog_bar=True, logger=True)
+        if self.finding_lr:
+            # Manually track the loss for the lr finder
+            self.log("train_loss", loss, on_step=False, on_epoch=True, prog_bar=True, logger=True)
```
Comment (ourownstory) [inline neuralprophet/time_net.py:806]: pass log_args

File: neuralprophet/time_net.py [hunk 7]
```diff
@@ -761,20 +762,21 @@ def loss_func(self, inputs, predicted, targets):
         loss = None
         # Compute loss. no reduction.
         loss = self.config_train.loss_func(predicted, targets)
-        # Weigh newer samples more.
-        loss = loss * self._get_time_based_sample_weight(t=inputs["time"][:, self.n_lags :])
+        if self.config_train.newer_samples_weight > 1.0:
+            # Weigh newer samples more.
+            loss = loss * self._get_time_based_sample_weight(t=inputs["time"][:, self.n_lags :])
```
Comment (ourownstory) [inline neuralprophet/time_net.py:767]: simplify to only pass first forecast target

File: neuralprophet/time_net.py [hunk 8]
```diff
@@ -861,48 +869,65 @@ def predict_step(self, batch, batch_idx, dataloader_idx=0):
...
+            )
+        else:
+            lr_scheduler = self.config_train.scheduler(
+                optimizer,
+                **self.config_train.scheduler_args,
+            )
 
         return {"optimizer": optimizer, "lr_scheduler": lr_scheduler}
 
     def _get_time_based_sample_weight(self, t):
```
Comment (ourownstory) [inline neuralprophet/time_net.py:904]: simplify to compute based only on first forecast target

File: neuralprophet/forecaster.py [hunk 9]
```diff
@@ -2773,84 +2796,68 @@ def _train(
...
 
-        # Init the model, if not continue from checkpoint
-        if continue_training:
-            raise NotImplementedError(
-                "Continuing training from checkpoint is not implemented yet. This feature is planned for one of the \
-                    upcoming releases."
-            )
-        else:
-            self.model = self._init_model()
+        self._init_model()
```
Comment (ourownstory) [inline neuralprophet/forecaster.py:2807]: if not self.fitted

File: neuralprophet/forecaster.py [hunk 10]
```diff
@@ -2773,84 +2796,68 @@ def _train(
...
             progress_bar_enabled=progress_bar_enabled,
             metrics_enabled=metrics_enabled,
             checkpointing_enabled=checkpointing_enabled,
-            num_batches_per_epoch=len(train_loader),
+            num_batches_per_epoch=batches_per_epoch,
             deterministic=deterministic,
         )
 
+        # Find suitable learning rate
+        if not self.config_train.learning_rate:
```
Comment (ourownstory) [inline neuralprophet/forecaster.py:2825]: init separate model and dataset/train_load and trainer for the lr-finder

File: neuralprophet/time_net.py [hunk 11]
```diff
@@ -794,19 +796,25 @@ def training_step(self, batch, batch_idx):
...
+            self.log("train_loss", loss, on_step=False, on_epoch=True, prog_bar=True, logger=True)
+            self.log("reg_loss", reg_loss, on_step=False, on_epoch=True, prog_bar=True, logger=True)
 
         # Metrics
-        if self.metrics_enabled:
+        if self.metrics_enabled and not self.finding_lr:
             predicted_denorm = self.denormalize(predicted[:, :, 0])
             target_denorm = self.denormalize(targets.squeeze(dim=2))
             self.log_dict(self.metrics_train(predicted_denorm, target_denorm), **self.log_args)
             self.log("Loss", loss, **self.log_args)
```
Comment (ourownstory) [inline neuralprophet/time_net.py:814]: rename train_loss, etc

File: neuralprophet/time_net.py (hunk 6 above)
Comment (ourownstory) [reply neuralprophet/time_net.py:806]: check for lr-finder requirements

File: neuralprophet/time_net.py (hunk 7 above)
Comment (ourownstory) [reply neuralprophet/time_net.py:767]: will do another time

File: neuralprophet/time_net.py (hunk 6 above)
Comment (ourownstory) [reply neuralprophet/time_net.py:806]: implement better prgress/metrics logging/printing. 
e.g. also need to touch `time_net.__init__`
```
self.log_args = {
                "on_step": False,
                "on_epoch": True,
                "prog_bar": True,
                "batch_size": self.config_train.batch_size,
            }
```

File: neuralprophet/time_net.py (hunk 8 above)
Comment (ourownstory) [reply neuralprophet/time_net.py:904]: postponed
